nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,535 @@
+    }
+
+    // Working with data sources parameters
+    source_names_ = get_parameter("observation_sources").as_string_array();
+    for (std::string source_name : source_names_) {
+      nav2_util::declare_parameter_if_not_declared(
+        node, source_name + ".type",
+        rclcpp::ParameterValue("scan"));  // Laser scanner by default
+      const std::string source_type = get_parameter(source_name + ".type").as_string();
+
+      if (source_type == "scan") {
+        std::shared_ptr<Scan> s = std::make_shared<Scan>(
+          node, tf_buffer_, source_name, base_frame_id_,
+          transform_tolerance, max_time_shift_);
+
+        if (!s->init()) {
+          return false;
+        }
+
+        sources_.push_back(s);
+      } else {  // source_type == "pointcloud"
+        std::shared_ptr<PointCloud> p = std::make_shared<PointCloud>(
+          node, tf_buffer_, source_name, base_frame_id_,
+          transform_tolerance, max_time_shift_);
+
+        if (!p->init()) {
+          return false;
+        }
+
+        sources_.push_back(p);
+      }
+    }
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(get_logger(), "Error while getting parameters: %s", ex.what());
+    return false;
+  }
+
+  // Store polygons coordinates for a future re-use when publishing
+  std::vector<Point> poly;
+  for (std::shared_ptr<PolygonBase> polygon : polygons_) {
+    polygon->getPolygon(poly);
+    for (Point p : poly) {
+      geometry_msgs::msg::Point32 p_s;
+      p_s.x = p.x;
+      p_s.y = p.y;
+      p_s.z = 0.0;
+      polygons_pub_.points.push_back(p_s);
+    }
+  }
+
+  return true;
+}
+
+void CollisionMonitorNode::workerMain()
+{
+  // Do nothing if main worker in non-active state
+  // or if velocity is not received yet
+  if (!getWorkerActive() || !velocityValid()) {
+    return;
+  }
+
+  // Get latest valid velocity for further usage
+  Velocity velocity = getVelocity();
+  // Points array collected from different data sources in a robot base frame
+  std::vector<Point> collision_points;
+  rclcpp::Time curr_time = this->now();
+
+  // Fill collision_points array from different data sources
+  for (std::shared_ptr<SourceBase> source : sources_) {
+    source->getData(collision_points, curr_time, velocity);
+  }
+
+  // ToDo: Add nearest points bounds estimation by using all polygons and current velocity
+  // and then cut-off of extra points out of calculated bounds
+
+  // By default - there is no action
+  Action ra{DO_NOTHING,{0.0, 0.0, 0.0}};
+  // Used for APPROACH model to define "nearest" obstacle
+  double min_collision_time = std::numeric_limits<double>::max();
+  // Current robot operation velocity: maximum possible one
+  Velocity max_vel = getCmdVelIn();
+
+  for (std::shared_ptr<PolygonBase> polygon : polygons_) {
+    const ActionType at = polygon->getActionType();
+    if (ra.action_type == STOP) {
+      // If robot already should stop, do nothing
+      break;
+    }
+
+    if (at == STOP || at == SLOWDOWN) {
```

Comment:
Break the stop / slowdown behavior into 1 function (since they're related actions) and break the approach into another function (since that's a standalone capability)

Reply:
[Done](https://github.com/ros-planning/navigation2/pull/2982/files#diff-1f10ee4d90224cdf0df5406c16550451120bdf02367ca835551eb99a2b061faaR133-R158)